library/psa_crypto.c

```diff
@@ -2818,7 +2932,28 @@ static psa_status_t psa_mac_setup( psa_mac_operation_t *operation,
 
     if( truncated == 0 )
     {
-        /* The "normal" case: untruncated algorithm. Nothing to do. */
+        /* The "normal" case: untruncated algorithm. Re-validate the
+         * key policy with explicit MAC length set in the algorithm
+         * when the algorithm policy is at-least-this-length to catch
+         * a corner case due to the default MAC length being unknown
+         * at key loading time. */
+        if( PSA_ALG_IS_MAC( slot->attr.policy.alg ) &&
```

Comment:
I really don't like doing policy checks in an operation function. While it isn't a blocker in itself, and the code isn't buggy directly because of this, it carries several risks:

* The general structure of the code is harder to understand.
* The operation code doesn't understand the complexities of policies (or won't be updated if policies become more compex).
* The policy check may be bypassed in some driver configurations (not the case now, but could slip in a refactoring).
* The policy check may be accidentally bypassed in the future, e.g. when one-shot MAC functions are implemented.

I would consider building more knowledge of MAC algorithms into the policy checks to be a lesser evil.

Reply:
And do you have a suggestion for *how* to do that, before we can get an indication of the default length? The issue here being that the default tag length is encoded as '0' for MAC algorithms, but as the actual default length in AEAD. The default size for MAC further depends on the exact cipher, so it depends on the key properties in addition to the algorithm specifier.

The other issue would be that using mbedtls_cipher_xxx operations to detect the underlying block cipher size won't be available anymore soon (when hardware accelerator drivers are plugged in and software fallback turned off).

So the only feasible way around would be to implement a support function in `psa_crypto.c`, which sets the algorithm value to be a truncated-MAC with the truncated length set to the default length in accordance with the key that is being used. That would still be logic that gets exercised from within `mac_setup`, since I don't think it makes sense to make `psa_key_policy_permits` take in the full key properties.

Reply:
>  don't think it makes sense to make psa_key_policy_permits take in the full key properties.

Passing the key type to policy functions (both `psa_key_policy_permits` and `psa_key_policy_algorithm_intersection`, they need to be consistent) seems to me like a lesser evil, compared to accessing the policy in an operation function.

Another possible approach could be to reject the problematic case in `psa_validate_key_attributes`: if `alg` or `alg2` has a policy of the form at-least-this-length-MAC, and the minimum length is more than the default length (calculated from the algorithm and the key type), reject the key as having an invalid policy.

Reply:
> `psa_validate_key_attributes`

On second thoughts, no, that's a bad idea. We currently accept keys that have a type that's incompatible with every usage algorithm permitted by their policy. This is not baked in stone: the PSA specification is deliberately agnostic about whether to accept such keys. But we should be consistent, and not just reject a few special cases.

Reply:
Good thing I didn't spend the last two hours going that route, then, but rather went the route of `policy_permits` and `policy_algorithm_intersection`. Updated now.